File: targets/watch/Shapes.swift
import SwiftUI

/// The same outlines as the phone app, sampled in a unit box centered on 0.5,
/// starting at 12 o'clock and running clockwise.
enum Outline {
  static let steps = 144

  static func points(for shape: String) -> [CGPoint] {
    switch shape {
    case "squircle": return superellipse(0.47, 4)
    case "square": return superellipse(0.46, 10)
    case "hexagon": return polygon(6, 0.47)
    case "star": return star(5, 0.48, 0.24)
    case "heart": return heart()
    case "blob": return polar { a in 0.42 + 0.04 * sin(3 * a + 0.6) + 0.03 * sin(5 * a + 2.1) }
    case "ripples": return wavy
    default: return superellipse(0.47, 2)
    }
  }

  /// The icon's wavy ring.
  static let wavy: [CGPoint] = polar { a in 0.45 + 0.03 * sin(8 * a + 0.4) }

  static func path(_ pts: [CGPoint], in rect: CGRect) -> Path {
    var p = Path()
    guard let first = pts.first else { return p }
    func map(_ pt: CGPoint) -> CGPoint {
      CGPoint(x: rect.minX + pt.x * rect.width, y: rect.minY + pt.y * rect.height)
    }
    p.move(to: map(first))
    for pt in pts.dropFirst() { p.addLine(to: map(pt)) }
    p.closeSubpath()
    return p
  }

  private static func polar(_ radius: (Double) -> Double) -> [CGPoint] {
    (0..<steps).map { i in
      let a = -Double.pi / 2 + Double(i) / Double(steps) * Double.pi * 2
      let r = radius(a)
      return CGPoint(x: 0.5 + r * cos(a), y: 0.5 + r * sin(a))
    }
  }

  private static func superellipse(_ a: Double, _ n: Double) -> [CGPoint] {
    polar { angle in
      let c = abs(cos(angle)), s = abs(sin(angle))
      return a / pow(pow(c, n) + pow(s, n), 1 / n)
    }
  }

  private static func polygon(_ sides: Int, _ radius: Double) -> [CGPoint] {
    (0..<sides).map { i in
      let a = -Double.pi / 2 + Double(i) * 2 * Double.pi / Double(sides)
      return CGPoint(x: 0.5 + radius * cos(a), y: 0.5 + radius * sin(a))
    }
  }

  private static func star(_ points: Int, _ outer: Double, _ inner: Double) -> [CGPoint] {
    (0..<(points * 2)).map { i in
      let r = i % 2 == 0 ? outer : inner
      let a = -Double.pi / 2 + Double(i) * Double.pi / Double(points)
      return CGPoint(x: 0.5 + r * cos(a), y: 0.5 + r * sin(a))
    }
  }

  private static func heart() -> [CGPoint] {
    var raw: [CGPoint] = []
    for i in 0..<steps {
      let t = Double(i) / Double(steps) * Double.pi * 2
      let x = 16 * pow(sin(t), 3)
      let y = -(13 * cos(t) - 5 * cos(2 * t) - 2 * cos(3 * t) - cos(4 * t))
      raw.append(CGPoint(x: x, y: y))
    }
    let xs = raw.map(\.x), ys = raw.map(\.y)
    let w = xs.max()! - xs.min()!, h = ys.max()! - ys.min()!
    let k = 0.94 / max(w, h)
    let cx = (xs.max()! + xs.min()!) / 2, cy = (ys.max()! + ys.min()!) / 2
    let pts = raw.map { CGPoint(x: 0.5 + ($0.x - cx) * k, y: 0.5 + ($0.y - cy) * k) }
    return [pts[0]] + pts.dropFirst().reversed()
  }
}

extension Color {
  /// "#RRGGBB" to Color. Anything else is a mid blue.
  init(hex: String) {
    var h = hex.trimmingCharacters(in: .whitespaces)
    if h.hasPrefix("#") { h.removeFirst() }
    if h.count == 3 { h = h.map { "\($0)\($0)" }.joined() }
    guard h.count == 6, let n = UInt32(h, radix: 16) else {
      self = Color(red: 0.17, green: 0.4, blue: 1)
      return
    }
    self.init(
      red: Double((n >> 16) & 255) / 255,
      green: Double((n >> 8) & 255) / 255,
      blue: Double(n & 255) / 255
    )
  }

  func lighter(_ amount: Double) -> Color {
    Color(hexMix(self, toward: (1, 1, 1), amount: amount))
  }

  func darker(_ amount: Double) -> Color {
    Color(hexMix(self, toward: (0, 0, 0), amount: amount))
  }
}

private func hexMix(_ c: Color, toward t: (Double, Double, Double), amount: Double) -> Color {
  #if canImport(UIKit)
  let ui = UIColor(c)
  var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
  ui.getRed(&r, green: &g, blue: &b, alpha: &a)
  return Color(
    red: Double(r) + (t.0 - Double(r)) * amount,
    green: Double(g) + (t.1 - Double(g)) * amount,
    blue: Double(b) + (t.2 - Double(b)) * amount
  )
  #else
  return c
  #endif
}

File: src/button/shapes.ts
import type { Shape } from "../types";

/**
 * Every shape is a closed polyline in a 100 x 100 box, sampled densely enough
 * to read as a smooth curve. Working from points (not SVG arcs and curves)
 * means one path can drive the button, the ripples, and the hold timer, and
 * the timer can trace the outline because the path length is known.
 *
 * All outlines start at 12 o'clock and run clockwise, so a timer that fills
 * along the path starts at the top for every shape.
 */
const SIZE = 100;
const CENTER = SIZE / 2;
const STEPS = 180;

type Point = [number, number];

/** Sample a polar function r(angle) into points, starting at the top, clockwise. */
function polar(radius: (angle: number) => number): Point[] {
  const pts: Point[] = [];
  for (let i = 0; i < STEPS; i++) {
    const a = -Math.PI / 2 + (i / STEPS) * Math.PI * 2;
    const r = radius(a);
    pts.push([CENTER + r * Math.cos(a), CENTER + r * Math.sin(a)]);
  }
  return pts;
}

/** Superellipse |x/a|^n + |y/a|^n = 1: n = 2 is a circle, 4 a squircle, 10 a rounded square. */
function superellipse(a: number, n: number): Point[] {
  return polar((angle) => {
    const c = Math.abs(Math.cos(angle));
    const s = Math.abs(Math.sin(angle));
    return a / Math.pow(Math.pow(c, n) + Math.pow(s, n), 1 / n);
  });
}

function polygon(sides: number, radius: number): Point[] {
  const pts: Point[] = [];
  for (let i = 0; i < sides; i++) {
    const a = -Math.PI / 2 + (i * 2 * Math.PI) / sides;
    pts.push([CENTER + radius * Math.cos(a), CENTER + radius * Math.sin(a)]);
  }
  return pts;
}

function star(points: number, outer: number, inner: number): Point[] {
  const pts: Point[] = [];
  for (let i = 0; i < points * 2; i++) {
    const r = i % 2 === 0 ? outer : inner;
    const a = -Math.PI / 2 + (i * Math.PI) / points;
    pts.push([CENTER + r * Math.cos(a), CENTER + r * Math.sin(a)]);
  }
  return pts;
}

/** The classic parametric heart, scaled to the box, top notch at 12 o'clock. */
function heart(): Point[] {
  const raw: Point[] = [];
  for (let i = 0; i < STEPS; i++) {
    const t = (i / STEPS) * Math.PI * 2;
    const x = 16 * Math.pow(Math.sin(t), 3);
    const y = -(13 * Math.cos(t) - 5 * Math.cos(2 * t) - 2 * Math.cos(3 * t) - Math.cos(4 * t));
    raw.push([x, y]);
  }
  // Fit to the box with a small margin.
  const xs = raw.map((p) => p[0]);
  const ys = raw.map((p) => p[1]);
  const w = Math.max(...xs) - Math.min(...xs);
  const h = Math.max(...ys) - Math.min(...ys);
  const k = 94 / Math.max(w, h);
  const cx = (Math.max(...xs) + Math.min(...xs)) / 2;
  const cy = (Math.max(...ys) + Math.min(...ys)) / 2;
  const pts = raw.map<Point>(([x, y]) => [CENTER + (x - cx) * k, CENTER + (y - cy) * k]);
  // t = 0 is the top notch; the parametric heart runs counterclockwise in screen
  // space, so reverse it to run clockwise like the other shapes.
  return [pts[0], ...pts.slice(1).reverse()];
}

/** The icon's wavy ring: a circle whose radius rises and falls `waves` times. */
export function wavyCircle(
  radius: number,
  waves: number,
  amplitude: number,
  phase = 0.4,
): Point[] {
  return polar((a) => radius + amplitude * Math.sin(waves * a + phase));
}

function toPath(pts: Point[]): string {
  return `${pts.map(([x, y], i) => `${i === 0 ? "M" : "L"}${x.toFixed(2)},${y.toFixed(2)}`).join(" ")} Z`;
}

function length(pts: Point[]): number {
  let total = 0;
  for (let i = 0; i < pts.length; i++) {
    const [x1, y1] = pts[i];
    const [x2, y2] = pts[(i + 1) % pts.length];
    total += Math.hypot(x2 - x1, y2 - y1);
  }
  return total;
}

export const SHAPE_POINTS: Record<Shape, Point[]> = {
  circle: superellipse(47, 2),
  squircle: superellipse(47, 4),
  square: superellipse(46, 10),
  hexagon: polygon(6, 47),
  star: star(5, 48, 24),
  heart: heart(),
  blob: polar((a) => 42 + 4 * Math.sin(3 * a + 0.6) + 3 * Math.sin(5 * a + 2.1)),
};

/** The icon's wavy ring, used for ripples and the Original mode rings. */
export const WAVY_PATH = toPath(wavyCircle(45, 8, 3));

export const SHAPE_PATHS: Record<Shape, string> = Object.fromEntries(
  (Object.keys(SHAPE_POINTS) as Shape[]).map((k) => [k, toPath(SHAPE_POINTS[k])]),
) as Record<Shape, string>;

/** Outline length in box units, for tracing the outline with a dash. */
export const SHAPE_LENGTHS: Record<Shape, number> = Object.fromEntries(
  (Object.keys(SHAPE_POINTS) as Shape[]).map((k) => [k, length(SHAPE_POINTS[k])]),
) as Record<Shape, number>;

/** How a ripple is outlined: the button's own shape, the icon's wavy ring, or a plain circle. */
export type RippleShape = "match" | "wavy" | "round";

export function ripplePath(shape: Shape, rippleShape: RippleShape): string {
  if (rippleShape === "wavy") return WAVY_PATH;
  if (rippleShape === "round") return SHAPE_PATHS.circle;
  return SHAPE_PATHS[shape];
}

export const SHAPE_VIEWBOX = `0 0 ${SIZE} ${SIZE}`;
/** Margin around the 100 box so glow copies and ripples are not clipped. */
export const GLOW_MARGIN = 30;
export const GLOW_VIEWBOX = `${-GLOW_MARGIN} ${-GLOW_MARGIN} ${SIZE + GLOW_MARGIN * 2} ${SIZE + GLOW_MARGIN * 2}`;
/** Multiply a button size by this to get the glow box size. */
export const GLOW_BOX_RATIO = (SIZE + GLOW_MARGIN * 2) / SIZE;
/** SVG transform that scales a path about the center of the 100 box. */
export function scaleAboutCenter(scale: number): string {
  return `translate(${CENTER} ${CENTER}) scale(${scale}) translate(${-CENTER} ${-CENTER})`;
}
